src/io/ioexternalbus.cpp

```diff
@@ -28,10 +45,201 @@ QString IOExternalBus::description() const {
-    return 0;
+    uint32_t rvalue = 0;
+
+    if (tcpSocket->isOpen()) {
+        while (skt_use)
+            QApplication::processEvents();  // TODO only need in rvss, change to mutex
+        skt_use = 1;
+        uint32_t payload = htonl(offset);
+        if (send_cmd(VB_PREAD, 4, (const char*)&payload) < 0) {
+            skt_use = 0;
+            return 0;
+        }
+
+        cmd_header_t cmd_header;
+
+        if (recv_cmd(&cmd_header) < 0) {
+            skt_use = 0;
+            return 0;
+        }
+
+        if (cmd_header.payload_size) {
+            uint32_t* payload = new uint32_t[2];
+            if (recv_payload((char*)payload, 8) < 0) {
+                skt_use = 0;
+                return 0;
+            }
+            for (uint32_t i = 0; i < 2; i++) {
+                payload[i] = ntohl(payload[i]);
+            }
+            dprintf("Read addr[%x] = %x \n", payload[0], payload[1]);
+            rvalue = payload[1];
+            delete[] payload;
+        } else {
+            skt_use = 0;
+            printf("read error [%lx] (%x) msg=%i pyaload_size=%i !!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!\n", offset, size,
+                   cmd_header.msg_type, cmd_header.payload_size);
+            tcpSocket->close();
+        }
+        skt_use = 0;
+    }
+
+    dprintf("===============>ioRead [%lx] = %x (%x)\n", offset, rvalue, size);
+    return rvalue;
+}
+
+void IOExternalBus::ioWrite(AInt offset, VInt value, unsigned size) {
+    dprintf("===============>ioWrite [%lx] = %lx (%x)\n", offset, value, size);
+    if (tcpSocket->isOpen()) {
+        while (skt_use)
+            QApplication::processEvents();  // TODO only need in rvss, change to mutex
+        skt_use = 1;
+        uint32_t payload[2];
+        payload[0] = htonl(offset);
+        payload[1] = htonl(value);
+        if (send_cmd(VB_PWRITE, 8, (const char*)payload) < 0) {
+            skt_use = 0;
+            return;
+        }
+
+        cmd_header_t cmd_header;
+
+        if (recv_cmd(&cmd_header) < 0) {
+            skt_use = 0;
+            return;
+        }
+
+        if (cmd_header.msg_type == VB_PWRITE) {
+            dprintf("write ok\n");
+        } else {
+            skt_use = 0;
+            printf("write error !!!!!!!!!!!!!!!!!!!!!!\n");
+            tcpSocket->close();
+        }
+        skt_use = 0;
+    }
+}
+
+void IOExternalBus::parameterChanged(unsigned paramId) { /* no parameters */
+}
+
+void IOExternalBus::connectButtonTriggered() {
+    if (m_ui->connectButton->text().contains("Connect")) {
+        tcpSocket->abort();
+        if (tcpSocket->connectToHost(m_ui->address->text().toStdString().c_str(), 7890)) {
+            if (send_cmd(VB_PINFO) < 0)
+                return;
+
+            cmd_header_t cmd_header;
+
+            if (recv_cmd(&cmd_header) < 0)
+                return;
+
+            if (cmd_header.payload_size) {
+                char* buff = new char[cmd_header.payload_size + 1];
```

Comment:
In modern C++, it's generally seen as code smell to have `new`/`delete` calls in the code unless it's strictly necessary (heap memory managers etc.) - why ? since there should be some level of abstraction that handles interaction with the dynamic memory allocation system. 
Again, switching to a QByteArray here would address this issue.

Reply:
I previously preferred the buff whit type char * to be able to see the values with the GDB. I changed to the QByteArray to encapsulate the dynamic memory allocation as you have indicated but now it is no longer possible to visualize the content with the GDB. Is there any way to do the GDB to view the QByteArray content?

Reply:
Hmm, to get an equivalent output, maybe storing the `QByteArray::data()` pointer somewhere gives insight into the array.
```c++
QByteArray arr = ...
char* data = arr.data();
```
Since i mainly use QtCreator when developing, there's the added benefit of introspection into most Qt containers (https://wiki.qt.io/IDE_Debug_Helpers).